### PR TITLE
[Backend] Implement background job for querying a specific keyword

### LINKED ISSUE #11: [Backend] As a user, I can query all the uploaded keywords from CSV file
Query all the keywords from the uploaded CSV file.

## Acceptance criteria
- [x] Handling user background job
- [ ] Save the result to database

### LINKED ISSUE #41: [Backend] Parse the search result
Implement a background job for querying specific keywords.

## Acceptance criteria
- [ ] Parse the HTML result and get the following data on the Google first result page:

```
1. Number of AdWords advertisers in the top position.
2. The total number of AdWords advertisers on the page.
3. URLs of the AdWords advertisers in the top position.
4. Number of the non-AdWords results on the page.
5. URLs of the non-AdWords results on the page.
6. Total number of links (all of them) on the page
```

### PR DESCRIPTION
**Pull request by byhbt**
Issue: 
- https://github.com/byhbt/kwtool/issues/9
- https://github.com/byhbt/kwtool/issues/41

## What happened

- [x] Implement background job for crawl Google search result
- [x] Parse the HTML response and get the required metrics.
- [x] Add `ExVCR` library for request testing.
- [x] Change the `async` of feature tests to `false` to avoid some random failed test.
- [x] Upgrade bootstrap to `v5`

## Insight

n/a
 
## Proof Of Work

Screen Shot 2022-01-17 at 1 40 23 PM

The test should pass.

### CODE REVIEW

File: lib/kwtool/crawler/keywords.ex [hunk 1]
```diff
@@ -36,16 +31,54 @@ defmodule Kwtool.Crawler.Keywords do
...
     |> Repo.one()
   end
 
+  def find_by_id!(keyword_id) do
+    Keyword
+    |> where(id: ^keyword_id)
+    |> Repo.one!()
+  end
+
+  def add_crawl_result(%Keyword{id: keyword_id}, crawler_result_attrs) do
```
Comment (rosle) [inline lib/kwtool/crawler/keywords.ex]: How about calling this `create_keyword_result`

File: test/kwtool/crawler/keywords_test.exs [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
       created_user = insert(:user)
 
       keyword_file = %Plug.Upload{
         content_type: "text/csv",
-        path: "test/support/fixtures/keywords.csv"
+        path: "test/support/fixtures/3-keywords.csv"
       }
 
-      assert {:ok, :file_is_proccessed} = Keywords.save_keywords_list(keyword_file, created_user)
+      assert {:ok, :file_is_processed} = Keywords.save_keywords_list(keyword_file, created_user)
```
Comment (rosle) [inline test/kwtool/crawler/keywords_test.exs:18]:
```suggestion
      assert Keywords.save_keywords_list(keyword_file, created_user) == {:ok, :file_is_processed}
```

File: test/kwtool/crawler/keywords_test.exs [hunk 3]
```diff
@@ -97,22 +105,64 @@ defmodule Kwtool.KeywordsTest do
     end
   end
 
-  describe "get_keyword_by_user/2" do
-    test "returns a keyword for a given user" do
-      created_user = insert(:user)
-      user_keyword = insert(:keyword, user: created_user)
-      keyword = Keywords.get_keyword_by_user(created_user, user_keyword.id)
+  describe "find_keyword_by_user/2" do
+    test "given the user and the keyword Id, returns a keyword with result" do
```
Comment (rosle) [inline test/kwtool/crawler/keywords_test.exs]:
```suggestion
    test "given the user and the keyword Id, returns a keyword with results" do
```

File: lib/kwtool/crawler/providers/google/crawler.ex [hunk 4]
```diff
@@ -0,0 +1,18 @@
+defmodule Kwtool.Providers.Google.Crawler do
```
Comment (rosle) [inline lib/kwtool/crawler/providers/google/crawler.ex]: The module name and the folder structure does not match. Should we update it? 💡

File: lib/kwtool/crawler/schemas/keyword.ex [hunk 5]
```diff
@@ -18,6 +19,8 @@ defmodule Kwtool.Crawler.Schemas.Keyword do
 
     belongs_to :user, User
 
+    has_many :keyword_results, KeywordResult
```
Comment (rosle) [inline lib/kwtool/crawler/schemas/keyword.ex:22]: Just wondering, when it can have multiple results? When the user search the same keyword again? 💭

Comment (byhbt) [reply lib/kwtool/crawler/schemas/keyword.ex:22]: yup, my idea about this feature is the user can see how the search result changed over time. but for now we only have one time. 😅

File: test/kwtool/crawler/keywords_test.exs (hunk 3 above)
Comment (byhbt) [reply test/kwtool/crawler/keywords_test.exs]: Updated in https://github.com/byhbt/kwtool/pull/44/commits/b1c39cfe7f5cfc78c042a456072e35f53b8aa354

File: test/kwtool/crawler/keywords_test.exs (hunk 2 above)
Comment (byhbt) [reply test/kwtool/crawler/keywords_test.exs:18]: Updated in https://github.com/byhbt/kwtool/pull/44/commits/b1c39cfe7f5cfc78c042a456072e35f53b8aa354

File: lib/kwtool/crawler/providers/google/crawler.ex (hunk 4 above)
Comment (byhbt) [reply lib/kwtool/crawler/providers/google/crawler.ex]: yeah, added `Keyword` to the namespace here:  https://github.com/byhbt/kwtool/pull/44/commits/6671c350d8aa598a283db074cf1511ff34d83fe9

Comment (byhbt) [reply lib/kwtool/crawler/providers/google/crawler.ex]: yeah, added `Keyword` to the namespace here:  https://github.com/byhbt/kwtool/pull/44/commits/6671c350d8aa598a283db074cf1511ff34d83fe9

File: lib/kwtool/crawler/keywords.ex (hunk 1 above)
Comment (byhbt) [reply lib/kwtool/crawler/keywords.ex]: Renamed the method here: https://github.com/byhbt/kwtool/commit/6671c350d8aa598a283db074cf1511ff34d83fe9

**Review by rosle — APPROVED** [review]
🚀

File: lib/kwtool/crawler/keywords.ex [hunk 6]
```diff
@@ -36,16 +31,54 @@ defmodule Kwtool.Crawler.Keywords do
...
+
+  def mark_as_finished(%Keyword{} = keyword) do
+    keyword
+    |> Keyword.finished_changeset()
+    |> Repo.update()
+  end
+
+  defp process_keyword_list(keyword_list, user_id) do
+    Enum.each(keyword_list, fn keyword ->
+      {:ok, %Keyword{id: keyword_id}} =
```
Comment (hanam1ni) [inline lib/kwtool/crawler/keywords.ex]: Should this handle when fail to create keyword?

File: lib/kwtool_web/templates/keyword/_keyword-result.html.eex [hunk 7]
```diff
@@ -0,0 +1,46 @@
+<%= if length(@keyword_results) > 0 do %>
```
Comment (hanam1ni) [inline lib/kwtool_web/templates/keyword/_keyword-result.html.eex]: I think using the function would make the condition easier to read
```suggestion
<%= if Enum.any?(@keyword_results) do %>
```

File: lib/kwtool_web/templates/keyword/_keyword-result.html.eex [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        <th scope="col">All Links Count</th>
+        <th scope="col">Inserted At</th>
+      </tr>
+    </thead>
+    <tbody>
+      <%= for keyword_result <- @keyword_results do %>
+        <tr>
+          <td><%= keyword_result.top_ads_count %></td>
+          <td>
+            <%= if length(keyword_result.top_ads_urls) > 0 do %>
```
Comment (hanam1ni) [inline lib/kwtool_web/templates/keyword/_keyword-result.html.eex]: Let's use `Enum.any?` here

Comment (byhbt) [reply lib/kwtool_web/templates/keyword/_keyword-result.html.eex]: Updated in https://github.com/byhbt/kwtool/pull/44/commits/213a5dfa7b88ebb6f5e12f03c5b2ea849a9b56ca

File: lib/kwtool_web/templates/keyword/_keyword-result.html.eex (hunk 7 above)
Comment (byhbt) [reply lib/kwtool_web/templates/keyword/_keyword-result.html.eex]: Updated in https://github.com/byhbt/kwtool/commit/026f5369772dd70e47d30f40105a139adb9a1ad7

File: lib/kwtool_web/templates/keyword/_keyword-result.html.eex (hunk 8 above)
Comment (byhbt) [reply lib/kwtool_web/templates/keyword/_keyword-result.html.eex]: Updated in https://github.com/byhbt/kwtool/commit/026f5369772dd70e47d30f40105a139adb9a1ad7

File: lib/kwtool/crawler/keywords.ex (hunk 6 above)
Comment (byhbt) [reply lib/kwtool/crawler/keywords.ex]: Updated in https://github.com/byhbt/kwtool/pull/44/commits/8397dc85de9d2e75b4014c26d781f0d5d5b03763

**Review by hanam1ni — APPROVED** [review]

**Pull request merged**